include/podio/DataSource.h

```diff
@@ -0,0 +1,179 @@
+  ///
+  void Initialize() override;
+
+  ///
+  /// @brief Retrieve from podio::DataSource a set of ranges of entries that
+  ///        can be processed concurrently.
+  ///
+  std::vector<std::pair<ULong64_t, ULong64_t>> GetEntryRanges() override;
+
+  ///
+  /// @brief Inform podio::DataSource that a certain thread is about to start
+  ///        working on a certain range of entries.
+  ///
+  void InitSlot(unsigned int slot, ULong64_t firstEntry) override;
+
+  ///
+  /// @brief Inform podio::DataSource that a certain thread is about to start
+  ///        working on a certain entry.
+  ///
+  bool SetEntry(unsigned int slot, ULong64_t entry) override;
+
+  ///
+  /// @brief Inform podio::DataSource that a certain thread finished working
+  ///        on a certain range of entries.
+  ///
+  void FinalizeSlot(unsigned int slot) override;
+
+  ///
+  /// @brief Inform podio::DataSource that an event-loop finished.
+  ///
+  void Finalize() override;
+
+  ///
+  /// @brief Returns a reference to the collection of the dataset's column
+  ///        names
+  ///
+  const std::vector<std::string>& GetColumnNames() const override;
+
+  ///
+  /// @brief Checks if the dataset has a certain column.
+  ///
+  bool HasColumn(std::string_view columnName) const override;
+
+  ///
+  /// @brief Type of a column as a string. Required for JITting.
+  ///
+  std::string GetTypeName(std::string_view columnName) const override;
+
+protected:
+  ///
+  /// @brief Type-erased vector of pointers to pointers to column
+  ///        values --- one per slot.
+  ///
+  std::vector<void*> GetColumnReadersImpl(std::string_view name, const std::type_info& typeInfo) override;
+
+  std::string AsString() override {
+    return "Podio data source";
+  };
+
+private:
+  /// Number of slots/threads
+  unsigned int m_nSlots = 1;
+
+  /// Input filename
+  std::vector<std::string> m_filePathList = {};
+
+  /// Total number of events
+  unsigned int m_nEvents = 0;
+
+  /// Ranges of events available to be processed
+  std::vector<std::pair<ULong64_t, ULong64_t>> m_rangesAvailable = {};
+
+  /// Ranges of events available ever created
+  std::vector<std::pair<ULong64_t, ULong64_t>> m_rangesAll = {};
+
+  /// Column names
+  std::vector<std::string> m_columnNames{};
+
+  /// Column types
+  std::vector<std::string> m_columnTypes = {};
+
+  /// Collections, m_Collections[columnIndex][slotIndex]
+  std::vector<std::vector<const podio::CollectionBase*>> m_Collections = {};
+
+  /// Active collections
+  std::vector<unsigned int> m_activeCollections = {};
+
+  /// Root podio readers
+  std::vector<std::unique_ptr<podio::Reader>> m_podioReaders = {};
+
+  /// Podio frames
+  std::vector<std::unique_ptr<podio::Frame>> m_frames = {};
+
+  ///
+  /// @brief Setup input for the podio::DataSource.
+  ///
+  /// @param[in]  Number of events.
+  /// @return     void.
+  ///
+  void SetupInput(int nEvents);
+};
+
+///
+/// Not used.
```

Comment:
If this is not used, can you explain why it is needed to be here?

Reply:
Not needed at all in the newer versions of ROOT, completely removed. This method was part of the old RDataSource interface, if I'm not mistaken.